pbft/pbft_persistent_state.hpp

```diff
@@ -0,0 +1,334 @@
+    // keys for each nested map, set, vector, etc.
+    // E.g. for map<int, persistent<std::string>> items, you would insert a string mapped to the integer 9 using:
+    // map.insert(std::make_pair(9, {storage, "value", "items", 9}); which would yield a storage key "items_00009".
+    //
+    // In the case of nested maps, multiple subkeys are needed to guarantee the uniqueness of the generated
+    // storage key. e.g. for map<int, map<char, persistent<string>>> nesty; you would insert an item in the 7 map,
+    // mapped to the char 'x' like so: nesty[7]['x'] = {storage, "value", "nesty", 'x', 7}; which would yield a
+    // storage key "nesty_x_00007".
+    //
+    // For consistency, it's recommended that in nested collections the sub-keys be ordered from inner to outer.
+    //
+    // Containers of persistent values MUST be initialized before use. Use one of the initialize methods below
+    // to load any existing contents of a container from storage before accessing the container, even if empty.
+    // The basic initialize methods take a lambda that should emplace elements into the container. There are also
+    // helper methods for initializing key/value-style containers such as maps.
+    //
+    // Each type used as a value or a key requires specialized to_string and from_string methods. Implementation are
+    // provided for std::string, uint64_t and a couple of types used by pbft. Unfortunately it's not possible to
+    // implement a generic *_string methods for tuples, or even for each tuple arity, as C++ doesn't allow
+    // specialization for (non-concrete) templated types without specializing the entire class template.
+    //
+    // In order to permanently remove a persistent variable it is necessary to call the destroy() method. This is
+    // because the variable's value is intended to persist after the object representing it is destructed.
+    // In the case of members of a collection (e.g. map, set), each member must be destroy()'d if you wish to erase
+    // it from storage. If you do not do this, it will re-appear with its previous value the next time the collection
+    // is initialized, or if an element with the same key is added later.
+    //
+    template <typename T>
+    class persistent : public persist_base
+    {
+    public:
+        // construct a persistent variable, retrieving its value from storage if it exists, otherwise
+        // initializing it with the provided default.
+        template <typename... K>
+        persistent(std::shared_ptr<bzn::storage_base> persist_storage, const T &default_value, const std::string& name, K... subkeys)
+            : storage(persist_storage), key(escape(name) + generate_key(subkeys...))
+        {
+            if (this->storage)
+            {
+                if (sizeof...(subkeys))
+                {
+                    if (initialized_containers.find(name) == initialized_containers.end())
```

Comment:
I suppose they could still be using an uninitialized map after a different alias of the map was initialized somewhere? I guess we're not likely to get it perfectly, though.